### PR TITLE
[global] Add Kubernetes 1.28 support

### PR DESCRIPTION
**Pull request by vitaliy-sn**
## Description

Add Kubernetes 1.28 support.

## Checklist
1.28:
- [x] AWS
- [x] Azure
- [x] GCP
- [x] OpenStack
- [x] vSphere
- [x] Yandex-cloud
- [x] Static
- [x] multimaster
- [x] update from 1.27
- [x] ingress
- [x] monitoring
- [x] grafana dashboards
- [x] upmeter
- [x] kubernetes certified

## Test results

<details>

```shell
root@test-master-0:~# kubectl get no -o wide
NAME                                                STATUS   ROLES                  AGE   VERSION   INTERNAL-IP   EXTERNAL-IP     OS-IMAGE             KERNEL-VERSION    CONTAINER-RUNTIME
test-master-0                      Ready    control-plane,master   27m   v1.28.2   10.0.0.2      34.32.243.173   Ubuntu 18.04.3 LTS   4.15.0-1042-gcp   containerd://1.6.24
test-master-1                      Ready    control-plane,master   21m   v1.28.2   10.0.0.3      35.204.163.35   Ubuntu 18.04.3 LTS   4.15.0-1042-gcp   containerd://1.6.24
test-master-2                      Ready    control-plane,master   21m   v1.28.2   10.0.0.4      34.90.180.154   Ubuntu 18.04.3 LTS   4.15.0-1042-gcp   containerd://1.6.24
test-worker-7e02228e-5886d-7gtxw   Ready    worker                 15m   v1.28.2   10.0.0.5      35.204.196.39   Ubuntu 18.04.3 LTS   4.15.0-1042-gcp   containerd://1.6.24
```

```shell
root@test-master-0:~# date
Mon Oct  2 16:26:39 UTC 2023
root@test-master-0:~# kubectl get no -o wide
NAME                               STATUS   ROLES                  AGE   VERSION   INTERNAL-IP   EXTERNAL-IP     OS-IMAGE             KERNEL-VERSION    CONTAINER-RUNTIME
test-master-0                      Ready    control-plane,master   27m   v1.27.6   10.0.0.2      34.90.43.26     Ubuntu 18.04.3 LTS   4.15.0-1042-gcp   containerd://1.6.24
test-worker-c02b8427-bcf8b-h46bj   Ready    worker                 20m   v1.27.6   10.0.0.3      35.204.196.39   Ubuntu 18.04.3 LTS   4.15.0-1042-gcp   containerd://1.6.24
root@test-master-0:~# date
Mon Oct  2 16:38:43 UTC 2023
root@test-master-0:~# kubectl get no -o wide
NAME                               STATUS   ROLES                  AGE   VERSION   INTERNAL-IP   EXTERNAL-IP     OS-IMAGE             KERNEL-VERSION    CONTAINER-RUNTIME
test-master-0                      Ready    control-plane,master   39m   v1.28.2   10.0.0.2      34.90.43.26     Ubuntu 18.04.3 LTS   4.15.0-1042-gcp   containerd://1.6.24
test-worker-c02b8427-bcf8b-h46bj   Ready    worker                 32m   v1.28.2   10.0.0.3      35.204.196.39   Ubuntu 18.04.3 LTS   4.15.0-1042-gcp   containerd://1.6.24
root@test-master-0:~# kubectl version
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.2
```

image
image

</details>

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: candi
type: feature
summary: Add Kubernetes 1.28 support.
impact: All control plane components will be restarted.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/static'...

**Comment by vitaliy-sn** [comment]
/e2e/abort static;Static;containerd;1.28 6336412897 6336412897-1-con-1-28 /sys/deckhouse-oss/install:pr5894 astra@192.168.199.91

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/gcp'...

**Comment by vitaliy-sn** [comment]
/e2e/abort gcp;WithoutNAT;containerd;1.28 6337307112 6337307112-1-con-1-28 /sys/deckhouse-oss/install:pr5894 user@34.89.226.26

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/azure'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/azure'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/azure'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/vsphere'...

**Review by yalosev — DISMISSED** [review]

File: candi/docs/README.md [hunk 1]
```diff
@@ -28,7 +28,7 @@ clusterType: Static
 podSubnetCIDR: 10.111.0.0/16
 podSubnetNodeCIDRPrefix: "24"
 serviceSubnetCIDR: 10.222.0.0/16
-kubernetesVersion: 1.23
+kubernetesVersion: 1.24
```
[the remark this answers is not in the document]
Comment (RomanenkoDenys) [reply candi/docs/README.md:31]: i think we should use automatic version here (1.25 for now)

File: candi/version_map.yml [hunk 2]
```diff
@@ -15,27 +15,6 @@ bashible: &bashible
...
-    csi:
-      openstack: v1.23.1
-      provisioner: v3.4.1
-      attacher: v4.2.0
-      resizer: v1.7.0
-      registrar: v2.7.0
-      snapshotter: v6.2.1
-      livenessprobe: v2.9.0
   '1.24':
     status: available
```
Comment (z9r5) [inline candi/version_map.yml:19]:
```suggestion
    status: end-of-life
```

@nabokihms Do we need to mark K8s 1.24 as EOL?

File: candi/docs/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (vitaliy-sn) [reply candi/docs/README.md]: Changed to 1.25

File: candi/openapi/cluster_configuration.yaml [hunk 3]
```diff
@@ -21,7 +21,7 @@ apiVersions:
       podSubnetNodeCIDRPrefix: "24"
       podSubnetCIDR: 10.244.0.0/16
       serviceSubnetCIDR: 192.168.0.0/16
-      kubernetesVersion: "1.23"
+      kubernetesVersion: "1.24"
```
[the remark this answers is not in the document]
Comment (vitaliy-sn) [reply candi/openapi/cluster_configuration.yaml]: Changed to 1.25

File: candi/control-plane-kubeadm/config.yaml.tpl [hunk 4]
```diff
@@ -2,14 +2,18 @@
 RotateKubeletServerCertificate default is true, but CIS becnhmark wants it to be explicitly enabled
 https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
 */}}
-{{- $featureGates := list "EndpointSliceTerminatingCondition=true" "TopologyAwareHints=true" "RotateKubeletServerCertificate=true" | join "," }}
+{{- $featureGates := list "TopologyAwareHints=true" "RotateKubeletServerCertificate=true" | join "," }}
+{{- if semverCompare "< 1.23" .clusterConfiguration.kubernetesVersion }}
+    {{- $featureGates = list $featureGates "EphemeralContainers=true" | join "," }}
+{{- end }}
```
Comment (RomanenkoDenys) [inline candi/control-plane-kubeadm/config.yaml.tpl:8]: we cannot have feature gates for 1.23, i think rebase is needed

File: ee/modules/030-cloud-provider-vsphere/images/cloud-controller-manager/patches/1.28/README.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
+## Patches
+
+## 002-network-name-discovery-logic.patch
+
+### Kubernetes versions prior to 1.22
+This patch adds InternalNetworkNames and  ExternalNetworkNames parameters to VirtualCenter configuration section. 
+Both of them we use to explicitly define what networks are external/internal to properly provide IP Addresses in the status of Node objects.
+
```
Comment (RomanenkoDenys) [inline ee/modules/030-cloud-provider-vsphere/images/cloud-controller-manager/patches/1.28/README.md:8]: i think we should remove this section from all readme's

Comment (vitaliy-sn) [reply ee/modules/030-cloud-provider-vsphere/images/cloud-controller-manager/patches/1.28/README.md]: It's not really important, but I deleted it

File: candi/control-plane-kubeadm/config.yaml.tpl [hunk 6]
```diff
@@ -2,14 +2,18 @@
...
 {{- if semverCompare "< 1.25" .clusterConfiguration.kubernetesVersion }}
     {{- $featureGates = list $featureGates "CustomResourceValidationExpressions=true" | join "," }}
 {{- end }}
 {{- if semverCompare "< 1.27" .clusterConfiguration.kubernetesVersion }}
     {{- $featureGates = list $featureGates "DaemonSetUpdateSurge=true" | join "," }}
 {{- end }}
 {{- if semverCompare "< 1.28" .clusterConfiguration.kubernetesVersion }}
+    {{- $featureGates = list $featureGates "EndpointSliceTerminatingCondition=true" | join "," }}
```
[the remark this answers is not in the document]
Comment (vitaliy-sn) [reply candi/control-plane-kubeadm/config.yaml.tpl:13]: Already done, maybe you didn't refresh the page.

File: candi/control-plane-kubeadm/config.yaml.tpl (hunk 4 above)
Comment (vitaliy-sn) [reply candi/control-plane-kubeadm/config.yaml.tpl]: deleted

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/aws'...

**Review by RomanenkoDenys — COMMENTED** [review]
some comments

**Comment by vitaliy-sn** [comment]
/e2e/abort aws;WithoutNAT;containerd;1.28 6496071660 6496071660-1-con-1-28 /sys/deckhouse-oss/install:pr5894 centos@3.79.214.245

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @vitaliy-sn. I've started the workflow for label 'e2e/run/aws'...

**Review by name212 — APPROVED** [review]

**Review by RomanenkoDenys — APPROVED** [review]
LGTM

**Pull request merged**